segmentstore/server/src/test/java/io/pravega/segmentstore/server/logs/DurableLogTests.java

```diff
@@ -1133,10 +1137,12 @@ public void testTruncateWithoutRecovery() {
                 // We were not able to do a full truncation before. Do one now, since we are guaranteed to have a new DataFrame available.
                 MetadataCheckpointOperation lastCheckpoint = new MetadataCheckpointOperation();
                 durableLog.add(lastCheckpoint, TIMEOUT).join();
+                awaitLastOperationAdded(durableLog, metadata);
```

Comment:
The PR description says:

> it was added just after the truncation executed

but here we are waiting before the truncation. What am I missing?

Reply:
The pipeline works in these steps:
1. The "operation" is queued up.
2. It is processed by the OP.
3. The OP acks it (after persisting into Tier 1).
4. The OP adds it to in-memory structures.

We proceed with the test after step 3, and step 4 executes concurrently in the background. It's possible that we initiate and complete the truncation after 3 and before 4 is completed, thus leading to this problem.

Reply:
That's not what I'm asking, let me add more detail to see if it clarifies. The PR description says:

> it was added just after the truncation executed

which is precisely what I see in lines 1099 and 1100. I see a different order in lines 1140 and 1141, and I'm wondering why the truncate in line 1141 isn't before the await in line 1140.

Reply:
Sorry, I had a confusing description. I reworded that sentence to make it clear that the operation would asynchronously end up in the log after we issue a truncation. We add it to the log **before**, then it is ack-ed, then we do a truncation, and due to some thread scheduling, the initial operation would only be added to the internal log after the truncation.